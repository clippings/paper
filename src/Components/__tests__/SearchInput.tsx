import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { SearchInput, SUBMIT_TYPES } from '@src';
jest.useFakeTimers();
jest.mock('lodash', () => ({
  debounce: jest.fn().mockImplementation((callback, delay) => {
    let timer: any = null;
    const resetTimer = () => {
      if (timer) {
        clearTimeout(timer);
      }
      timer = null;
    };
    const wrapped = (...args) => {
      resetTimer();
      timer = setTimeout(() => {
        callback(...args);
      }, delay);
    };
    wrapped.cancel = () => {
      resetTimer();
    };
    return wrapped;
  }),
}));
const onChange = jest.fn();
describe('Search input', () => {
  const searchInput = <SearchInput placeholder="Search ..." onChange={onChange} />;
  test('match snapshot', () => {
    const { container } = render(searchInput);
    expect(container).toMatchSnapshot();
  });
  test('onChange is triggered', () => {
    onChange.mockClear();
    const { container } = render(searchInput);
    const input = container.querySelector('input');
    if (input !== null) {
      fireEvent.keyUp(input, {
        target: {
          value: 'Test',
        },
      });
    }

    jest.runOnlyPendingTimers();
    expect(onChange).toBeCalled();
    expect(onChange).toBeCalledWith('Test');
  });
  test('onChange is triggered once', () => {
    onChange.mockClear();
    const debouncedInput = (
      <SearchInput placeholder="Search ..." onChange={onChange} debounceRate={500} />
    );
    const { container } = render(debouncedInput);
    const input = container.querySelector('input');
    if (input !== null) {
      fireEvent.keyUp(input, {
        target: {
          value: 'Test',
        },
      });
      fireEvent.keyUp(input, {
        target: {
          value: 'Test Test',
        },
      });
    }
    jest.runOnlyPendingTimers();
    expect(onChange).toBeCalledTimes(1);
    expect(onChange).toBeCalledWith('Test Test');
  });
  test('onChange is triggered on enter', () => {
    onChange.mockClear();
    const searchOnEnterInput = (
      <SearchInput placeholder="Search ..." onChange={onChange} submitOn={SUBMIT_TYPES.ENTER} />
    );
    const { container } = render(searchOnEnterInput);
    const input = container.querySelector('input');
    if (input !== null) {
      fireEvent.keyUp(input, {
        target: {
          value: 'Test',
        },
      });
      fireEvent.keyUp(input, {
        key: 'Enter',
        code: 'Enter',
      });
    }
    jest.runOnlyPendingTimers();
    expect(onChange).toBeCalledTimes(1);
    expect(onChange).toBeCalledWith('Test');
  });
});
