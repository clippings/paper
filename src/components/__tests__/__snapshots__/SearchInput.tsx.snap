// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search input match snapshot 1`] = `
<div>
  <div
    class="p-f-search"
  >
    <input
      class="p-f-search__input"
      placeholder="Search ..."
      type="text"
      value=""
    />
    <svg
      fill="#8f8f8f"
      height="16"
      preserveAspectRatio="xMidYMid meet"
      viewBox="0 0 14 14"
      width="16"
    >
      <path
        clip-rule="evenodd"
        d="M9.16437 9.87147C8.30243 10.5768 7.20063 11 6 11C3.23858 11 1 8.76142 1 6C1 3.23858 3.23858 1 6 1C8.76142 1 11 3.23858 11 6C11 7.20064 10.5768 8.30243 9.87147 9.16437L12.8366 12.1295L12.1295 12.8366L9.16437 9.87147ZM10 6C10 8.20914 8.20914 10 6 10C3.79086 10 2 8.20914 2 6C2 3.79086 3.79086 2 6 2C8.20914 2 10 3.79086 10 6Z"
        fill-rule="evenodd"
      />
    </svg>
  </div>
</div>
`;
